README.md
# RGordon
Localised bottlenecks for server transport layer protocol analysis.  

## Dependancies:
```quick_dependancies.sh```
----------------------------

###Configuration:
```cd RScripts```
```./configure.sh```

### Runnning tests:
-----------------
### Trial runs with dummy Server
1. Launch dummy server  
```$cd dummyServer```  
```$npm start```  
2. Run the Scripts  
```node RScripts/dummyTest.js```  

## Actual Runs
1. Run the following command  
```$ node RScripts/worker.js```  
2. To configure as well as run the worker
```$ cd RScripts```
```$ ./start.sh```


history
       start_emudrop: '100000',
       startRTT: '22',
       endRTT: '22',
       chances_left: '3',
       trials: '12',
       cwnd: '27',
       sigma_cwnd: '731' } ] }
python3 calculate.pyhttps://ju.taobao.com/tg/today_items.htm?type=0&spm=1.1.10.0 12 731 27 22 100000
Entering 22
calculate.py done for 22
child process exited with code 0
{ cwnd: '29',
  sigma_cwnd: '760',
  last_rtt_done: '22',
  url:
   'https://ju.taobao.com/tg/today_items.htm?type=0&spm=1.1.10.0',
  emuDrop: '100000',
  viewpoint: 'Singapore' }
{ message: 'Data updated successfully' }
Asking for a job
{ message: 'JOB',
  data:
   [ { url:
        'https://accounts.google.com/ServiceLogin?hl=en&passive=true&continue=https://www.google.com/%3Fgws_rd%3Dssl',
       viewpoint: 'singapore',
       start_emudrop: '100000',
       startRTT: '25',
       endRTT: '25',
       chances_left: '5',
       trials: '12',
       cwnd: '104',
       sigma_cwnd: '1612' } ] }
python3 calculate.pyhttps://accounts.google.com/ServiceLogin?hl=en&passive=true&continue=https://www.google.com/%3Fgws_rd%3Dssl 12 1612 104 25 100000
Entering 25
calculate.py done for 25
child process exited with code 0
{ cwnd: '104',
  sigma_cwnd: '1716',
  last_rtt_done: '25',
  url:
   'https://accounts.google.com/ServiceLogin?hl=en&passive=true&continue=https://www.google.com/%3Fgws_rd%3Dssl',
  emuDrop: '1612',
  viewpoint: 'Singapore' }
{ message: 'Data updated successfully' }
Asking for a job
{ message: 'JOB',
  data:
   [ { url: 'https://www.youtube.com/',
       viewpoint: 'singapore',
       start_emudrop: '100000',
       startRTT: '24',
       endRTT: '24',
       chances_left: '4',
       trials: '12',
       cwnd: '104',
       sigma_cwnd: '1722' } ] }
python3 calculate.pyhttps://www.youtube.com/ 12 1722 104 24 100000
Entering 24
calculate.py done for 24
child process exited with code 0
{ cwnd: '106',
  sigma_cwnd: '1828',
  last_rtt_done: '24',
  url: 'https://www.youtube.com/',
  emuDrop: '1722',
  viewpoint: 'Singapore' }
{ message: 'Data updated successfully' }
Asking for a job
{ message: 'JOB',
  data:
   [ { url:
        'https://www.baidu.com/s?wd=%E5%A5%94%E9%A9%B0suv&rsf=62040005&rsp=8&f=1&oq=%E5%A5%94%E9%A9%B0%E5%85%A8%E7%90%83%E6%80%BB%E8%A3%81%E9%80%80%E4%BC%91&tn=SE_fengyunbangBD_x3kuz7ef&ie=utf-8&usm=2&rsv_pq=c61eee1c000041e4&rsv_t=d45dpRU%2B8rg%2FMIcsRj6i%2FXLrRh9XVv8UWZAOCO84rlxyGMUKMWEDMd1jt4TI6FjNG9gWbMk2YVdsqdl7QpRsVA&rqlang=cn&rs_src=0&rsv_pq=c61eee1c000041e4&rsv_t=d45dpRU%2B8rg%2FMIcsRj6i%2FXLrRh9XVv8UWZAOCO84rlxyGMUKMWEDMd1jt4TI6FjNG9gWbMk2YVdsqdl7QpRsVA',
       viewpoint: 'singapore',
       start_emudrop: '100000',
       startRTT: '23',
       endRTT: '23',
       chances_left: '3',
       trials: '12',
       cwnd: '102',
       sigma_cwnd: '1209' } ] }
python3 calculate.pyhttps://www.baidu.com/s?wd=%E5%A5%94%E9%A9%B0suv&rsf=62040005&rsp=8&f=1&oq=%E5%A5%94%E9%A9%B0%E5%85%A8%E7%90%83%E6%80%BB%E8%A3%81%E9%80%80%E4%BC%91&tn=SE_fengyunbangBD_x3kuz7ef&ie=utf-8&usm=2&rsv_pq=c61eee1c000041e4&rsv_t=d45dpRU%2B8rg%2FMIcsRj6i%2FXLrRh9XVv8UWZAOCO84rlxyGMUKMWEDMd1jt4TI6FjNG9gWbMk2YVdsqdl7QpRsVA&rqlang=cn&rs_src=0&rsv_pq=c61eee1c000041e4&rsv_t=d45dpRU%2B8rg%2FMIcsRj6i%2FXLrRh9XVv8UWZAOCO84rlxyGMUKMWEDMd1jt4TI6FjNG9gWbMk2YVdsqdl7QpRsVA 12 1209 102 23 100000
Entering 23
calculate.py done for 23
child process exited with code 0
{ last_error: 'error',
  last_rtt_done: '23',
  url:
   'https://www.baidu.com/s?wd=%E5%A5%94%E9%A9%B0suv&rsf=62040005&rsp=8&f=1&oq=%E5%A5%94%E9%A9%B0%E5%85%A8%E7%90%83%E6%80%BB%E8%A3%81%E9%80%80%E4%BC%91&tn=SE_fengyunbangBD_x3kuz7ef&ie=utf-8&usm=2&rsv_pq=c61eee1c000041e4&rsv_t=d45dpRU%2B8rg%2FMIcsRj6i%2FXLrRh9XVv8UWZAOCO84rlxyGMUKMWEDMd1jt4TI6FjNG9gWbMk2YVdsqdl7QpRsVA&rqlang=cn&rs_src=0&rsv_pq=c61eee1c000041e4&rsv_t=d45dpRU%2B8rg%2FMIcsRj6i%2FXLrRh9XVv8UWZAOCO84rlxyGMUKMWEDMd1jt4TI6FjNG9gWbMk2YVdsqdl7QpRsVA',
  chances_left: '2',
  viewpoint: 'Singapore' }
{ message: 'Data updated successfully' }
Asking for a job
{ message: 'JOB',
  data:
   [ { url: 'https://it.wikipedia.org/wiki/Stati_Uniti_d%27America',
       viewpoint: 'singapore',
       start_emudrop: '100000',
       startRTT: '25',
       endRTT: '25',
       chances_left: '5',
       trials: '12',
       cwnd: '107',
       sigma_cwnd: '1830' } ] }
python3 calculate.pyhttps://it.wikipedia.org/wiki/Stati_Uniti_d%27America 12 1830 107 25 100000
Entering 25
calculate.py done for 25
child process exited with code 0
{ cwnd: '105',
  sigma_cwnd: '1935',
  last_rtt_done: '25',
  url: 'https://it.wikipedia.org/wiki/Stati_Uniti_d%27America',
  emuDrop: '1830',
  viewpoint: 'Singapore' }
{ message: 'Data updated successfully' }
Asking for a job
{ message: 'JOB',
  data:
   [ { url:
        'https://ju.taobao.com/tg/today_items.htm?type=0&spm=1.1.10.0',
       viewpoint: 'singapore',
       start_emudrop: '100000',
       startRTT: '23',
       endRTT: '23',
       chances_left: '3',
       trials: '12',
       cwnd: '29',
       sigma_cwnd: '760' } ] }
python3 calculate.pyhttps://ju.taobao.com/tg/today_items.htm?type=0&spm=1.1.10.0 12 760 29 23 100000
Entering 23
calculate.py done for 23
child process exited with code 0
{ cwnd: '32',
  sigma_cwnd: '792',
  last_rtt_done: '23',
  url:
   'https://ju.taobao.com/tg/today_items.htm?type=0&spm=1.1.10.0',
  emuDrop: '100000',
  viewpoint: 'Singapore' }
{ message: 'Data updated successfully' }
Asking for a job
{ message: 'JOB',
  data:
   [ { url:
        'https://accounts.google.com/ServiceLogin?hl=en&passive=true&continue=https://www.google.com/%3Fgws_rd%3Dssl',
       viewpoint: 'singapore',
       start_emudrop: '100000',
       startRTT: '26',
       endRTT: '26',
       chances_left: '5',
       trials: '12',
       cwnd: '104',
       sigma_cwnd: '1716' } ] }
python3 calculate.pyhttps://accounts.google.com/ServiceLogin?hl=en&passive=true&continue=https://www.google.com/%3Fgws_rd%3Dssl 12 1716 104 26 100000
Entering 26
calculate.py done for 26
child process exited with code 0
{ cwnd: '104',
  sigma_cwnd: '1820',

RScripts/work.py
import json
import multiprocessing as mp
import subprocess
import requests
import sys
import re
fileName = sys.argv[1]
server=sys.argv[2]
threshold=80
defaultEmu=100000
json_file = open(fileName)
data = json.load(json_file)

def runJob(i,data):
    startRTT=int(data[i]['startRTT'])
    # print("startRTT %d" %(startRTT))
    endRTT=int((data[i]['endRTT']))
    # print("endRTT %d" %(endRTT))
    emuDrop = int((data[i]['start_emudrop']))
    # print("emu %d" %(emuDrop))
    chances_left=int((data[i]['chances_left']))
    # print("chances %d" %(chances_left))
    trials=int((data[i]['trials']))
    # print("trials %d" %(trials))
    cwnd=int((data[i]['cwnd']))
    # print("cwnd %d" %(cwnd))
    sigma_cwnd=int((data[i]['sigma_cwnd']))
    # print("scwnd %d" %(sigma_cwnd))
    url=(data[i]['url'])
    # print("url "+ url)
    viewPoint=data[i]['viewpoint']
    rnum=startRTT
    jobID=i
    rnum=startRTT

    targetURL=url
    response=None
    delayTime=50
    try:
        response = subprocess.check_output(
            ['ping', '-c', '1', url],
            stderr=subprocess.STDOUT,  # get all output
            universal_newlines=True  # return string not bytes
        )
    except subprocess.CalledProcessError:
        response = None

    if response == None:
        pingTime = -1
    else:
        pingTime = float(re.search('time=.*', response).group().replace(" ms", '')[5:])

    if int(pingTime/2) >= 50:
        delayTime = 1
    elif pingTime == -1:
        delayTime = 50
    else:
        delayTime = 50 - int(pingTime/2)


    for j in range(endRTT-startRTT+1):

        subprocess.call(["python3 calculate.py \""+url + "\" "+ str(trials)+ " "+str(sigma_cwnd )+ " "+str(cwnd )+ " "+str(rnum) +" "+ str(emuDrop)+" "+str(jobID)+" "+str(delayTime)],shell=True,executable='/bin/bash')

        infile="./RData"+str(jobID)+"/windows"+".csv"
        read=open(infile,'r')
        line=[int (x) for x in read.readline().split(' ')]
        values=line
        postData=''
        path=''
        toBreak=False
        if (values[1] == 0):
            toBreak=True
            chances_left=chances_left-1
            postData={'last_error':'error','last_rtt_done':str(rnum),'url':url,'chances_left':str(chances_left),'viewpoint':viewPoint}
            path='/api/worker/updateError'

        else:
            path='/api/worker/update'
            if(emuDrop==defaultEmu):
                if(values[1]>threshold):
                    emuDrop=sigma_cwnd
            cwnd=values[1]
            sigma_cwnd=values[0]
            postData={'cwnd':str(values[1]),'sigma_cwnd':str(values[0]),'last_rtt_done':str(values[2]),'url':url,'emudrop':str(emuDrop),'viewpoint':viewPoint,'max_trials':str(trials)}
        headers={'Content-type':'application/json','Accept':'text/plain'}
        requests.post(server+path,data=json.dumps(postData),headers=headers)
        if(toBreak):
            break
        rnum=rnum+1


if data['message']=='JOB':


    pool = mp.Pool(max(mp.cpu_count(), 5))
    print(len(data['data']))
    r=[pool.apply_async(runJob,args=[i,data['data']]) for i in range(len(data['data']))]
    p=[x.wait() for x in r]
    pool.close()
    # for i in range(len(data['data'])):
    #     runJob(i,data['data'])
    subprocess.call(["./clean.sh"],shell=True,executable='/bin/bash')
